File: python version/ANC.py
"""Single-channel FxLMS demonstration translated from ``ANC.m``."""

from __future__ import annotations

import argparse
from typing import Any

import matplotlib.pyplot as plt
import numpy as np
from scipy.signal import lfilter

from ANC_algorithm import ANC_algorithm
from demo_utils import matlab_fir1_bandpass, time_vector


def run_demo(
    *,
    Fs: int = 16_000,
    T: float = 30.0,
    wlen: int = 512,
    slen: int = 256,
    muw: float = 0.001,
    seed: int | None = None,
    plot: bool = True,
) -> dict[str, Any]:
    """Run the same synthetic single-channel setup as the MATLAB script."""

    t = time_vector(Fs, T)
    N = t.size
    rng = np.random.default_rng(seed)

    # Primary and secondary paths: MATLAB fir1(511, ...) and fir1(255, ...).
    Pri_path = matlab_fir1_bandpass(511, 20.0, 6000.0, Fs)
    Sec_path = matlab_fir1_bandpass(255, 20.0, 6000.0, Fs)
    if slen != Sec_path.size:
        raise ValueError(
            f"This demo creates a {Sec_path.size}-tap secondary path; "
            f"received slen={slen}."
        )

    # Broadband reference noise from 100 to 1000 Hz.
    noise = rng.standard_normal(N)
    fil = matlab_fir1_bandpass(127, 100.0, 1000.0, Fs)
    Ref = lfilter(fil, [1.0], noise)
    Dis = lfilter(Pri_path, [1.0], Ref)

    # SciPy has no direct equivalent of MATLAB dsp.FilteredXLMSFilter.  The
    # repository's own FxLMS class is therefore the canonical Python result.
    single_ANC = ANC_algorithm(wlen, slen, Sec_path, Dis, Ref)
    err, single_ANC = single_ANC.ANC_FxLMS(muw)

    if plot:
        sample_count = min(int(round(T * Fs)), N)
        plt.figure()
        plt.plot(t[:sample_count], Dis[:sample_count], label="Disturbance")
        plt.plot(t[:sample_count], err[:sample_count], label="Residual error signal")
        plt.title("Active Noise Control")
        plt.xlabel("Time (seconds)")
        plt.ylabel("Signal value")
        plt.legend()
        plt.grid(True)
        plt.tight_layout()
        plt.show()

    return {
        "t": t,
        "Ref": Ref,
        "Dis": Dis,
        "err": err,
        "controller": single_ANC,
    }


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--duration", type=float, default=30.0, help="Duration in s.")
    parser.add_argument("--seed", type=int, default=None)
    parser.add_argument("--no-plot", action="store_true")
    return parser.parse_args()


if __name__ == "__main__":
    args = parse_args()
    run_demo(T=args.duration, seed=args.seed, plot=not args.no_plot)

File: MATLAB version/ANC.m
%% single channel ANC

clc;clear;close all;

set(groot,'defaultAxesTickLabelInterpreter','latex');


Fs = 16000; % sampling frequency
T  = 30;     % time (s)
t  = 0:1/Fs:T;
N  = length(t);

%% Primary path
p_low  = 20  ;
p_high = 6000 ;
Pri_path = fir1(511,[2*p_low/Fs 2*p_high/Fs]);

%% secondary path
s_low = 20 ;
s_high = 6000;
Sec_path = fir1(255,[2*s_low/Fs 2*s_high/Fs]);

%% noise generation
noise = randn(N,1);  % random noise
low = 100;
high = 1000;
fil = fir1(127,[2*low/Fs 2*high/Fs]); % broadband 100 to 1,000Hz
Ref = filter(fil,1,noise);   % reference
Dis = filter(Pri_path,1,Ref);

%% system parameters
wlen = 512;  % control filter length
slen = 256;  % secondary path length
muw = 0.001; % step size

%% call matlab function https://www.mathworks.com/help/releases/R2025a/dsp/ref/dsp.filteredxlmsfilter-system-object.html?searchPort=54096#bva0bw8
fxlms = dsp.FilteredXLMSFilter(wlen, 'StepSize', muw, 'SecondaryPathCoefficients', Sec_path,'SecondaryPathEstimate',Sec_path);
[y,e] = fxlms(Ref,Dis); % y: control signal; e: error signal; e = Dis + filter(Sec_path,1,y)

figure;
plot(t(1:T*Fs),Dis(1:T*Fs));
hold on;
plot(t(1:T*Fs),e(1:T*Fs))
title('Active Noise Control','Interpreter','latex');
legend('Disturbance','Residual error signal','Interpreter','latex');
xlabel('Time Index'); 
ylabel('Signal Value');

%% FxLMS algorithm
single_ANC = ANC_algorithm(wlen,slen,Sec_path',Dis,Ref);
[err,single_ANC] = ANC_FxLMS(single_ANC,muw);

figure;
plot(t(1:T*Fs),Dis(1:T*Fs));
hold on;
plot(t(1:T*Fs),err(1:T*Fs))
title('Active Noise Control','Interpreter','latex');
legend('Disturbance','Residual error signal','Interpreter','latex');
xlabel('Time Index'); 
ylabel('Signal Value');

% compare control filter
figure;
plot(-fxlms.Coefficients);
hold on;
plot(single_ANC.Wc);
title('Coefficients comparsion','Interpreter','latex');
legend('Matlab function','FxLMS','Interpreter','latex');
xlabel('Tap length'); 
ylabel('Coefficients Value');
